File: LeetCode/Problems/Python/56. Merge Intervals.py
#My first try: heap
#Time: O(nlogn)
#Space: O(n)
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        heap = []
        for i in intervals:
            heapq.heappush(heap, (i[0],-1))
            heapq.heappush(heap, (i[1],1))
        
        conti = 0
        result = []
        start = 0
        end = 0
        while heap:
            curr = heapq.heappop(heap)
            conti+=curr[1]
            if conti == -1 and curr[1]==-1:
                start = curr[0]
            if conti == 0:
                end = curr[0]
                result.append([start,end])
                #print([start,end])
        
        return result



class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:

        if not intervals:
            return []
        
        intervals.sort()
        start = intervals[0][0]
        end = intervals[0][1]
        result = []
        
        
        for interval in intervals[1:]:
            if interval[0] > end:
                result.append([start,end])
                start = interval[0]
            end = max(end, interval[1])
            #else:
            #    end = interval[1]
        if not result or start > result[-1][1]:
            result.append([start,end])
        
        
        return result

File: LeetCode/Problems/C++/56. Merge Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if(!intervals.size()) return {};
        
        priority_queue<pair<int, int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        for(int i=0; i<intervals.size();i++){
            pq.push({intervals[i][0],-1});
            pq.push({intervals[i][1],1});
        }
        
        vector<vector<int>> result;
        int start = 0;
        int end = 0;
        int status = 0;
        while(!pq.empty()){
            pair<int, int> cur = pq.top();
            pq.pop();
            status += cur.second;

            if(status == -1 && cur.second == -1)
                start = cur.first;

            if(status == 0){
                end = cur.first;
                result.push_back({start,end});
            } 
        }      
        return result;
    }
};